calvin/runtime/south/plugins/transports/calvinfcm/twisted/twisted_transport.py

```diff
@@ -0,0 +1,278 @@
+from calvin.utilities import calvinconfig
+from twisted.protocols.basic import Int32StringReceiver
+
+from twisted.internet import protocol, reactor, ssl
+from twisted.names.srvconnect import SRVConnector
+from twisted.words.protocols.jabber import client
+from twisted.words.xish import xmlstream as xishxmlstream
+from twisted.words.protocols.jabber import xmlstream
+
+from twisted.words.protocols.jabber.jid import JID
+
+_log = calvinlogger.get_logger(__name__)
+
+_CONF = calvinconfig.get()
+
+def create_uri(sender_id, token):
+    return "calvinfcm://%s:%s" % (sender_id, token)
+
+class CalvinXMLStreamProtocol(CalvinCBClass, xmlstream.XmlStream):
+
+    def __init__(self, *args, **kvargs):
+        self._authenticator = args[0]
+        self._callbacks = args[1]
+        self._uri = args[2]
+        self._server = args[3]
+        xmlstream.XmlStream.__init__(self, self._authenticator)
+        CalvinCBClass.__init__(self, self._callbacks)
+        self.counter = 0
+        #self._callback_execute('set_proto', self)
+
+    def send(self, obj):
+        return super(CalvinXMLStreamProtocol, self).send(obj)
+
+    def sendCalvinMsg(self, to, data, msgType="payload"):
+        toSend ="<message id=\"\"><gcm xmlns=\"google:mobile:data\">%s</gcm></message>"
+        # TODO: Fix message id to something better...
+        
+        # Payload should be base64 encoded
+        # Add 4 byte data length before data
+        dataSize = len(data)
+        dlen = chr(dataSize >> 24 & 0xFF) + chr(dataSize >> 16 & 0xFF) + chr(dataSize >> 8 & 0xFF) + chr(dataSize & 0xFF) + data
+        data = base64.b64encode(dlen)
+        d = {
+                "to": to,
+                "message_id": "UPID%s%d" % (str(time.time()), self.counter),
+                "priority": "high",
+                "data": {
+                        "msg_type": msgType,
+                        "payload": data
+                    },
+                "time_to_live": 600                
+                }
+        self.counter = self.counter + 1
+        self.send(toSend % str(json.dumps(d)))
+
+    def dataReceived(self, data):
+        return super(CalvinXMLStreamProtocol, self).dataReceived(data)
+
+    def connectionLost(self, reason):
+        self._callback_execute('disconnected', reason)
+        _log.info("here we should remove all callbacks")
+        valid_names = self.callback_valid_names()
+        _log.info("Valid names: %s" % str(valid_names))
+        ids = []
+        for name in valid_names:
+            _log.info("Removing cbs for %s" % name)
+            cbs = self.get_callbacks_by_name(name)
+            _log.info("got cbs: %s" % cbs)
+            for cb_id in cbs:
+                ids.append(cb_id)
+        
+        for cb_id in ids:
+            self.callback_unregister(cb_id)
+
+        valid_names = self.callback_valid_names()
+        _log.info("Valid names should be none here: %s" % str(valid_names))
```

Comment:
I removed it instead. There is nothing we can do if it fails to remove a callback. I only used it for debugging.